### PR TITLE
🚧 Add pep8speaks bot basic configuration

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
Follow-up of #288, this PR adds `flake8` on Travis configs as pep8 checker. It is still a work in progress and is just a suggestion for testing.

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
[Count=1431](https://travis-ci.org/pyvista/pyvista/jobs/602273295#L2244) so far but I didn't introduce any rules. Let's try again after `max-line-length=88` (suggested in this [comment](https://github.com/pyvista/pyvista/issues/288#issuecomment-507626345)). It probably requires some tweaks after all.

File: setup.cfg [hunk 1]
```diff
@@ -5,3 +5,7 @@ all_files  = 1
 
 [upload_sphinx]
 upload-dir = docs/_build/html
+
+[flake8]
+exclude = __init__.py
+max-line-length = 88
```
Comment (agramfort) [inline setup.cfg]: MNE uses this:

https://github.com/mne-tools/mne-python/blob/master/setup.cfg#L19

Comment (GuillaumeFavelier) [reply setup.cfg]: E241: [Multiple spaces after ','](https://github.com/PyCQA/pycodestyle/issues/4)
E305: [Expected 2 blank lines after end of function or class](https://lintlyci.github.io/Flake8Rules/rules/E305.html)
W504: [Line break occurred after a binary operator](https://lintlyci.github.io/Flake8Rules/rules/W504.html)

Comment (GuillaumeFavelier) [reply setup.cfg]: I kinda like E305 so I would keep it but I agree with the rest. Thanks a lot for the suggestion @agramfort

**Comment by GuillaumeFavelier** [comment]
It's a matter of preference. Some examples of `setup.cfg`:
- [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/master/setup.cfg#L27-L29)
- [flask](https://github.com/pallets/flask/blob/master/setup.cfg#L22-L45)
- [django](https://github.com/django/django/blob/master/setup.cfg#L5-L8)

**Comment by GuillaumeFavelier** [comment]
[Count=1136] this time, it's a bit challenging. As expected the 3 modified jobs are not happy.

**Comment by GuillaumeFavelier** [comment]
This could maybe help #408 , what do you think @pyvista/developers ?

**Comment by banesullivan** [comment]
I think both @akaszynski and I are on board with setting up a code style checker like this and I am definitely in favor of adding an auto formatter like black (in accordance with #288). I wonder if we should set up black or autopep8 and run the formatter on the library (locally?) to automatically fix most of those errors? Only thing is that we may need to resolve as many other PRs/branches as possible or get everything to the same head so that we can easily propagate the formatting changes across branches (it’ll be a huge diff).

As for #408... I would really prefer to redo that PR so that it doesn’t have any stylistic changes, just feature changes/refactoring. It seems there are already too many conflicts on that diff to merge so we may need to clean up that work manually.

**Comment by GuillaumeFavelier** [comment]
Whether it is done automatically or not, once the library is formatted, the code style checkers configured properly should be enough to ensure consistency in coding style right?

> Only thing is that we may need to resolve as many other PRs/branches as possible or get everything to the same head so that we can easily propagate the formatting changes across branches (it’ll be a huge diff).

I agree, many PRs will require update but it will benefit greatly the future contributions.

> As for #408... I would really prefer to redo that PR so that it doesn’t have any stylistic changes.

+1 to separate somehow style and feature. It will simplify the review process too. Let me know if I can help.

**Comment by banesullivan** [comment]
> Whether it is done automatically or not, once the library is formatted, the code style checkers configured properly should be enough to ensure consistency in coding style right?

Yes, though one thing that bothers me about this is that I don't want a CI test to fail because of formatting/style - if that happens, it will discourage first-time contributors. Perhaps we continue using codacy ([Codacy Badge](https://www.codacy.com/manual/banesullivan/pyvista?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=pyvista/pyvista&amp;utm_campaign=Badge_Grade)) and have the core devs regularly run an autoformatter to make sure it stays consistent (i.e. myself and other volunteers set up a pre-commit hook to run Black). We may need to better configure codacy - I think I have it ignoring tons of pep codes.

 I don't want to put that burden/confusion on new contributors and I want to keep the barrier to entry for contributing to PyVista small.

**Comment by banesullivan** [comment]
I've caught up all PRs except #268 - that one looks like it has too many conflicts and it was stopped in a broken spot, so I'll likely have to completely restart that one.

**Comment by GuillaumeFavelier** [comment]
> Yes, though one thing that bothers me about this is that I don't want a CI test to fail because of formatting/style - if that happens, it will discourage first-time contributors.

I understand. Do you share this point of view @akaszynski ?

> Perhaps we continue using codacy and have the core devs regularly run an autoformatter to make sure it stays consistent (i.e. myself and other volunteers set up a pre-commit hook to run Black)

I don't know how to configure Codacy but I'm ready to learn. And I can set up a pre-commit hook too btw @banesullivan  👍  (once a formatter is decided).

Should we close this then?

**Comment by akaszynski** [comment]
@GuillaumeFavelier, I love PEP8, but I often go over 79 characters per line and I will cut corners sometimes as I see fit for the situation.  Even with emacs, `elpy`, `flycheck`, I still can't create perfect code, and it would drive me insane if a CI check failed because I wasn`t PEP8 compliant.

Ideally, the CI check would warn if someone is pushing code that isn't compliant.  Codacy seems like a happy medium as it lets us know when our code is becoming deviant.

**Comment by agramfort** [comment]
I would discourage you from using a code formatter. It can create ugly code
and you loose the occasion to learn. I think it's better to educate
contributors
to adopt contributing guidelines. You have also bots like
https://pep8speaks.com
that can help. All projects I contribute to have such linters in their CIs
with
some rules that are skipped depending on maintainers.

>

**Comment by akaszynski** [comment]
Agreed.  I've seen the output from `black` and it's less human readable than I'd like.  A report is better than a pass/fail or auto-formatter.  +1 for pep8speaks.com

**Comment by GuillaumeFavelier** [comment]
I just translated the `[flake8]` section of the `setup.cfg` file I modified into a `pep8speaks` configuration file. I'll let you add the bot to the project if you're happy with it.

Useful material: https://github.com/OrkoHunter/pep8speaks#configuration

**Comment by banesullivan** [comment]
We may want to add these flake8 plugins to prevent me from committing typos:

- https://pypi.org/project/flake8-spellcheck/
- https://pypi.org/project/flake8-docstrings/

**Comment by GuillaumeFavelier** [comment]
Apparently flake8 plugins are [not integrated yet in pep8speaks](https://github.com/OrkoHunter/pep8speaks/issues/56), it's on the [low priority section of the main board](https://github.com/OrkoHunter/pep8speaks/projects/1#card-18316019) for some time now.

**Comment by banesullivan** [comment]
That's fine then - I enabled them locally and that should be good enough

**Comment by banesullivan** [comment]
Should we close this since we have a routine configured to run codespell and pycodestyle on Azure builds now?

**Comment by akaszynski** [comment]
I really like `flake8`, and I've got a ton of warnings whenever I open up the source locally.  Perhaps we can reconsider this later; `pycodestyle` and `codespell` are good enough for the time being.  I vote close if @GuillaumeFavelier is fine with it.

**Comment by banesullivan** [comment]
Marked for future consideration

**Pull request closed without merging**